### PR TITLE
feat: support operation hooks

### PR DESCRIPTION
**Pull request by hacksparrow**
Added support for operation hooks. Addresses https://github.com/strongloop/loopback-next/issues/3952.

Signed-off-by: Hage Yaapa <hage.yaapa@in.ibm.com>

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/migration/models/operation-hooks.md [hunk 1]
```diff
@@ -6,8 +6,38 @@ sidebar: lb4_sidebar
...
+  typeof Product.prototype.id,
+  ProductRelations
+> {
+  constructor(dataSource: juggler.DataSource) {
+    super(Product, dataSource);
+  }
+
+  definePersistedModel(entityClass: typeof Product) {
+    const model = super.definePersistedModel(entityClass);
+    model.observe('before save', async ctx => {
```
Comment (raymondfeng) [inline docs/site/migration/models/operation-hooks.md]: We can access the model as `this.modelClass`. Why not leveraging that to set up observers instead of overriding `definedPersistedModel`?

**Comment by hacksparrow** [comment]
Needs https://github.com/strongloop/loopback-datasource-juggler/pull/1820 to land for tests to pass.

File: docs/site/migration/models/operation-hooks.md (hunk 1 above)
Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: Initially we were considering having an `observe` method on `DefaultCrudRepository`, which would call `this.modelClass.observe`. After some brainstorming, @bajtos suggested the possible problems with `observe` is not worth it. Eg: calling `observe()` multiple times and how to handle it. He had suggested another approach but it wasn't very DX-friendly, so we settled on the current one.

"You want to directly access the model? Then you will have to override the `definePersistedModel ` method."

Comment (bajtos) [reply docs/site/migration/models/operation-hooks.md]: @raymondfeng please check the description of the user story and the discussion in that GH issue for more context.

@hacksparrow please update PR description with a reference to the GH issue this is related to.

Comment (raymondfeng) [reply docs/site/migration/models/operation-hooks.md]: @hacksparrow We do support multiple observers for a given model class/operation - see https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/observer.js#L51

Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: Yes, that's understood. We were wondering how should LB4 respond when an operation hook is installed multiple times: a. Throw an error? b. Accept silently - how does the user know what happened?

Also, it is not impossible to call the `observe` method before the `modelClass` is defined. What happens in that case?

Since this is a temporary workaround, we didn't want to invest too much time handling various possible scenarios, edge cases etc. We can do that when we investigate the use of interceptor for the "real" implementation.

File: docs/site/migration/models/operation-hooks.md [hunk 2]
```diff
@@ -6,8 +6,38 @@ sidebar: lb4_sidebar
 permalink: /doc/en/lb4/migration-models-operation-hooks.html
 ---
 
-{% include note.html content="
-This is a placeholder page, the task of adding content is tracked by the
-following GitHub issue:
-[loopback-next#3952](https://github.com/strongloop/loopback-next/issues/3952)
-" %}
+Operation hooks are not supported in LoopBack 4 yet. It will be, eventually; you
```
Comment (dhmlau) [inline docs/site/migration/models/operation-hooks.md]: Are we really going to have operation hooks in LB4? Maybe reword it as:
`See the [Operation hooks for models/repositories spike](https://github.com/strongloop/loopback-next/issues/1919) for the progress`

Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: Updated as per suggestion.

File: docs/site/migration/models/operation-hooks.md [hunk 3]
```diff
@@ -6,8 +6,38 @@ sidebar: lb4_sidebar
...
+in LoopBack 4: override `DefaultCrudRepository`'s `definePersistedModel` method
+in the model's repository.
+
+The `definePersistedModel` method of `DefaultCrudRepository` returns a model
+instance on which you can apply the
+[LoopBack 3 operation hooks](https://loopback.io/doc/en/lb3/Operation-hooks.html).
+Make sure to return the model instance from the derived class.
+
+Here is an example of a repository implementing `definePersistedModel` and
+applying an operation hook on a model instance:
```
Comment (jannyHou) [inline docs/site/migration/models/operation-hooks.md]: nitpick: I remember the `observe` method is called by a model class, not instance, not sure is if it's still the case in the legacy juggler bridge.

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
```
Comment (agnes512) [inline packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: // Copyright IBM Corp. 2020

File: docs/site/migration/models/operation-hooks.md (hunk 3 above)
Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: Thanks! Corrected.

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+// This test shows the recommended way how to use @loopback/repository
+// together with existing connectors when building LoopBack applications
+describe('Operation hooks', () => {
+  let repo: ProductRepository;
+  beforeEach(givenProductRepository);
+
+  it('supports operation hooks', async () => {
+    const p = await repo.create({slug: 'pencil'});
+    expect(repo.callCounter).to.equal(2);
```
Comment (bajtos) [inline packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: When a check like this fails, it's hard to tell what exactly went wrong, because we don't know which hooks were called and which were not.

Also if somehow the hook registration logic misbehaves and registers the callback for wrong hooks (e.g. `access` instead of `before save`), then we won't catch such problem.

In the past, I have good experience with replacing a counter with an array of hook names called.

```suggestion
    expect(repo.hooksCalled).to.eql(['before save', 'after save']);
```

File: docs/site/migration/models/operation-hooks.md [hunk 6]
```diff
@@ -6,8 +6,39 @@ sidebar: lb4_sidebar
...
+
+  definePersistedModel(entityClass: typeof Product) {
+    const modelClass = super.definePersistedModel(entityClass);
+    modelClass.observe('before save', async ctx => {
+      console.log(`going to save ${ctx.Model.modelName}`);
+    });
+    return modelClass;
+  }
+}
+```
```
Comment (bajtos) [inline docs/site/migration/models/operation-hooks.md]: Please explain the problem of multiple registration of the same hook and how we are avoiding it by registering the hooks inside `definePersistedModel` method.

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts (hunk 5 above)
Comment (hacksparrow) [reply packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: 👍

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts (hunk 4 above)
Comment (hacksparrow) [reply packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: Updated

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts (hunk 5 above)
Comment (hacksparrow) [reply packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: Done

File: docs/site/migration/models/operation-hooks.md (hunk 6 above)
Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: Done

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 7]
```diff
@@ -5,7 +5,7 @@
 
 import {Getter} from '@loopback/context';
 import assert from 'assert';
-import legacy from 'loopback-datasource-juggler';
+import legacy, {ObserverMixin} from 'loopback-datasource-juggler';
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: I think we are intentionally keeping juggler types very separated from LB4 types. Take a look at lines 41-53, where we are re-exporting a subset of legacy types in `juggler` namespace.

Please make the following changes:

- Revert this line back.
- Re-export `legacy.ObserverMixin` in `juggler` namespace 
- Modify all places below using `ObserverMixin` to use `juggler.Mixin`

Please add a comment to all those modified places to mention that it's a temporary workaround and add a link to the GitHub issue tracking the follow-up task of improving juggler typings to correctly include ObserverMixin in `PersistedModelClass` API.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 8]
```diff
@@ -127,13 +127,12 @@ export class DefaultCrudRepository<
       `Entity ${entityClass.name} must have at least one id/pk property.`,
     );
 
-    this.modelClass = this.definePersistedModel(entityClass);
+    this.modelClass = this.ensurePersistedModel(entityClass);
   }
 
-  // Create an internal legacy Model attached to the datasource
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Please preserve this comment.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 9]
```diff
@@ -145,9 +144,19 @@ export class DefaultCrudRepository<
...
     if (model) {
       // The backing persisted model has been already defined.
-      return model as typeof juggler.PersistedModel;
+      return model as typeof juggler.PersistedModel & ObserverMixin;
     }
 
+    return this.definePersistedModel(entityClass);
+  }
+
+  // Create an internal legacy Model attached to the datasource
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Please replace this comment with a proper tsdoc-style one, explain LB4 app developers that they can override this method in their custom repository classes to tweak the legacy Model (e.g. register operation hooks).

File: packages/repository/src/repositories/legacy-juggler-bridge.ts (hunk 7 above)
Comment (hacksparrow) [reply packages/repository/src/repositories/legacy-juggler-bridge.ts]: 👍

File: packages/repository/src/repositories/legacy-juggler-bridge.ts (hunk 9 above)
Comment (hacksparrow) [reply packages/repository/src/repositories/legacy-juggler-bridge.ts]: 👍

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 10]
```diff
@@ -144,16 +148,23 @@ export class DefaultCrudRepository<
...
 
     return this.definePersistedModel(entityClass);
   }
 
-  // Create an internal legacy Model attached to the datasource
+  /**
+   * Create a model class and attach it to the datasource. This method can be
+   * overriden in sub-classes to acess methods and properties in the generated
+   * model class.
+   * @param entityClass - Legacy entity class
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: This is not correct. `entityClass` is the LB4 `Entity` constructor, e.g. `Product` or `User` inheriting from LB4 `Entity`.

The function returns the legacy persisted model class (e.g. `Product` inheriting from `juggler.PersistedModel`).

Please update this doc accordingly.

Comment (hacksparrow) [reply packages/repository/src/repositories/legacy-juggler-bridge.ts]: Updated to:
```ts
  /**
   * Creates a legacy persisted model class, attaches it to the datasource and
   * returns it. This method can be overriden in sub-classes to acess methods
   * and properties in the generated model class.
   * @param entityClass - LB4 Entity constructor
   */
```

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+describe('Operation hooks', () => {
+  let repo: ProductRepository;
+  beforeEach(givenProductRepository);
+
+  const beforeSave = 'before save';
+  const afterSave = 'after save';
+  const expectedArray = [beforeSave, afterSave];
+
+  it('supports operation hooks', async () => {
+    const p = await repo.create({slug: 'pencil'});
```
Comment (dhmlau) [inline packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: code linter complaining about `p` is not being used. 
Should change this to:
```
await repo.create({slug: 'pencil'});
```

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  > {
+    constructor(dataSource: juggler.DataSource) {
+      super(Product, dataSource);
+    }
+
+    hooksCalled: string[] = [];
+
+    definePersistedModel(entityClass: typeof Product) {
+      const modelClass = super.definePersistedModel(entityClass);
+      modelClass.observe(beforeSave, async ctx => {
```
Comment (dhmlau) [inline packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: code linter is complaining about `Promise returned in function argument where a void return was expected`. It might be because we're waiting for your changes in juggler to be released: https://github.com/strongloop/loopback-datasource-juggler/pull/1820.

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        this.hooksCalled.push(beforeSave);
+      });
+
+      modelClass.observe(afterSave, async ctx => {
```
Comment (dhmlau) [inline packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: code linter is complaining about `Promise returned in function argument where a void return was expected`. It might be because we're waiting for your changes in juggler to be released: https://github.com/strongloop/loopback-datasource-juggler/pull/1820.

File: packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts (hunk 11 above)
Comment (hacksparrow) [reply packages/repository/src/__tests__/acceptance/operation-hooks.acceptance.ts]: 👍

**Comment by hacksparrow** [comment]
@slnode test please

**Comment by hacksparrow** [comment]
@slnode test please

File: docs/site/migration/models/operation-hooks.md [hunk 14]
```diff
@@ -6,8 +6,46 @@ sidebar: lb4_sidebar
...
+
+Although possible, we are not providing an API which directly exposes the
+`observe` method of the model class. The current API makes the registration of
+operation hooks a process that is possible only after the model class is
```
Comment (bajtos) [inline docs/site/migration/models/operation-hooks.md]:
```suggestion
operation hooks a process that is possible only at the time when the model class is
```

Comment (hacksparrow) [reply docs/site/migration/models/operation-hooks.md]: 👍

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 15]
```diff
@@ -148,6 +148,21 @@ export class DefaultCrudRepository<
       return model as typeof juggler.PersistedModel;
     }
 
+    return this.definePersistedModel(entityClass);
```
Comment (raymondfeng) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Why don’t we emit an event when definePersistedModel is done so that listeners can be registered to set up operation hooks? Overriding a method is always more advanced for a lot of developers.

Comment (raymondfeng) [reply packages/repository/src/repositories/legacy-juggler-bridge.ts]: For the record, we briefly discussed if it's feasible to use events to set up operation hooks and decided to defer that.
